### DISCUSSION TITLE
Only show page when the data it uses is in the state, page renders before router, crashes app

### DISCUSSION THREAD

**Discussion by camondrey** [mobxjs/mobx#2945] [Q&A] [answered]
Hi all,

I have a setup where I have a number of pages, each of which has some conditions that need to be met for the page to be able to be shown. For example, I have a Project page that has a `projectId` passed to it, and if the associated project with that `id` isn't in the state, then the router should redirect to the Projects page.

Each page assumes that its conditions are met and should never be rendered if that is not the case. If the data the page you're on uses is deleted, the router should redirect (change the local state) to a different page so that the page with the missing data is not shown the moment we detect the data is gone.

The issue I'm facing is that when a project is deleted, and you're on the Project page, the Project page will notice that its data is missing before the router has a chance to redirect away from the page. This means that the whole app crashes. Both the router and the page are observers.

Since I'm accessing the project data that is used on the Project page in the router component when checking if the conditions are met, I thought that as a result the router, being a parent of the pages, would be rerendered first when the data changes. However, this is not the case and I've spent a huge amount of time trying to figure out a way for this to work, to no avail.

What am I doing wrong? What is the right way to create what is effectively an authentication component? I can share code showing how I'm accessing the project data, the structure of the state, etc, if that would help.

Thanks

**Comment by mweststrate** [comment]
Did you wrap your state updates in an action? If that is the case, observables that are used by parents and children will cause the parents to always render before any of their children (if they're observer)

**Reply by camondrey** [reply]
Indeed. To delete the project, I call an action in my store called `onDeleteProject` which finds the project instance by id and then calls its `destroy` action, which in turn removes the project from the list of projects in the store, by calling `store.projects.remove(this)`.

**Reply by camondrey** [reply]
Each condition specifies what type of entity we're after and its id. The router (parent) checks the conditions by going over each one and then gets that particular entity (e.g. `state.projects.find(project => project.id === id)`). If it is undefined, then that means the conditions failed, and we need to redirect by not showing the original page and instead show a different page.

**Comment by urugator** [comment]
> the router, being a parent of the pages, would be rerendered first when the data changes

If both parent and child component are invalidated at the same time (during the same action), the react should re-render parent first. May not be completely true for concurrent mode, dunno.

How do you perform the redirect? Do you check the conditions during router's render to shortcircuit Project page rendering and then again in effect to initiate the redirect?

Personally I would probably move the check and redirect to the action that deletes Project.

**Reply by camondrey** [reply]
What do you mean by invalidated? And also by concurrent mode?

I have a `page` observable in my state containing an `id` and `data`. If I want to navigate to a particular page, I use an action called `setPage` in my state which takes `id` and `data` and patches the existing instance of `Page`. Then, I use the value of `page.id` to determine what page to show.

Regarding moving the redirect to the action that deletes the project, that wouldn't solve the issue as the project could also be deleted by someone else, for example, so it's not enough to redirect away when deleting the project.

**Comment by mweststrate** [comment]
Note that action should be as far outward as possible, so that you don't
end up with two transactions instead of one. So typically the event handler
that cause your page to be deleted should be the action.

**Reply by camondrey** [reply]
The actions are in the state, not in an `runInAction` block in my components etc.

**Comment by camondrey** [comment]
Could the issue be that on the Project page I get the project data `project` and also access its `name` property, so I'm accessing `project.name`, whereas in the router I only get the `project` but don't use any of its properties. Maybe then since I access a specific property on the page, it will be rerendered first when the project is `undefined`?

**Comment by mweststrate** [comment, marked as answer]
1. please make sure your event handlers are state.
2. assuming that the project is removed from some collection, a component
reading that collection should re-render near the root

If that doesn't help it will be a lot more productive to set up a sandbox
showing the problem, as sharing vague directions on code we can't see or
run is really not very productive.

**Reply by camondrey** [reply]
Is it possible that the issue could be stemming from the way the actions in the store are called? I tried to make a small sandbox but every time it worked after removing our API. In our actual app we have document event listeners for each action. E.g. if I want to delete a project, I call an API handler and that in turn dispatches an event called `onDeleteProject` which triggers the `onDeleteProject` action in the state which deletes the project.

**Reply by camondrey** [reply]
It looks like this is exactly it - I initialised my state with some fake data and then when I add/delete projects, if I interact directly with the state it all works.

**Comment by camondrey** [comment]
After doing some more digging, I found that wrapping the action call in the API callback in a `runInAction` block solved the issue:

![image](https://user-images.githubusercontent.com/8189543/118978552-10101380-b9aa-11eb-8698-a2af447dddce.png)

In the code above `handlerId` is e.g `deleteProject` and `store` is an instance of my MobX state. So then my question boils down to this: Why do I need to use `runInAction` if the callback directly calls a single action in the state? I call `makeAutoObservable` in the constructor which should make class functions such as `onDeleteProject` actions automatically.

**Reply by mweststrate** [reply]
I can't look at `onDeleteProject`, so it is hard to answer your questions. But as said, it is the event handlers that should be action, so wrapping `action` around your `api.on` callback is the correct thing to do (`api.on(handlerId, action(data => { }))` is the slightly more convenient form). See also the docs on async actions.

**Reply by camondrey** [reply]
Thanks for the help, I got it to work.

**Reply by camondrey** [reply]
Interestingly, moving the `forEach` above outside the file where the MobX store class and instance were declared to the `index.js` of my app fixed it.
